ofrak_core/ofrak/service/data_service.py

```diff
@@ -234,32 +236,64 @@ def _apply_patches_to_root(
         ]
 
 
-# Helper classes
-@dataclass
-class _Waypoint:
-    offset: int
-    models_starting: Set[DataId]
-    models_ending: Set[DataId]
+T = TypeVar("T")
 
-    def is_empty(self) -> bool:
-        return not self.models_starting and not self.models_ending
+
+class _ShiftBreaksSortError(RuntimeError):
+    pass
+
+
+class _CompareFirstTuple(Tuple, Generic[T]):
+    """
+    Wrapper for tuple that ensures only the first item in the tuple is checked.
+    Necessary because bisect methods don't have a `key` function
+    Helpful for making sorted dictionary-like data structures
+    """
+
+    def __new__(cls, *args):
+        return super().__new__(cls, args)
+
+    def __lt__(self, other):
+        return self[0] < other[0]
+
+    def __eq__(self, other):
+        return self[0] == other[0]
+
+    @staticmethod
+    def bisect_left(grid, val: int) -> int:
+        return bisect_left(grid, _CompareFirstTuple(val, None))
+
+    @staticmethod
+    def bisect_right(grid, val: int) -> int:
+        return bisect_right(grid, _CompareFirstTuple(val, None))
+
+
+_GridYAxisT = List[_CompareFirstTuple[Set[DataId]]]
+_GridXAxisT = List[_CompareFirstTuple[_GridYAxisT]]
```

Comment:
The big problem with the containers in `sortedcontainers` is that they don't support shifting a range of keys in O(N) time. Specifically, if we resize a range (A, B), every "point" after B needs to be shifted (up or down, doesn't matter). The key thing is that the order doesn't change, so this could be just a O(N) operation, but with `SortedList` or `SortedDict` it becomes N remove-and-inserts operations, which are each O(N*logN), and so becomes O(N^2 * logN) time. This was a big time sink in an earlier iteration, and was why I sort of made a custom data structure.

The point about insertions is a good one though. We could use a binary tree and still have the O(N) key-shifting as described, with constant-time inserts and removals. I believe all of the random-access instances are really part of a "search for index of value and random-access that" which are easy to rewrite with a tree. 

However, it would be a bit complex to rewrite, and at the moment inserting into the data service is not a bottleneck for resource creation. So while strictly looking at the DataService it might be an obvious improvement, OFRAK workflows are unlikely to benefit from it. I will try profiling a couple cases that might be affected and think on it some more.

Reply:
This is a good point – I assume there is no convenient way to construct a sorted container from an already-sorted iterable, otherwise they would be a good fit. 

Though I think it is worth benchmarking if possible, I don't think adding trees should necessarily block the PR. But if you choose not to add them in, I think it makes sense to add a comment to somewhere in this file explaining that using them would be an easy optimization for later.